src/main/java/org/aksw/simba/squirrel/analyzer/mime/MimeTypeDetector.java

```diff
@@ -2,36 +2,42 @@
 
 import org.aksw.simba.squirrel.analyzer.impl.RDFAnalyzer;
 import org.apache.jena.riot.Lang;
+import org.apache.jena.riot.RDFLanguages;
 import org.slf4j.Logger;
 import org.slf4j.LoggerFactory;
 
 import java.io.File;
 import java.io.FileInputStream;
 import java.io.IOException;
+import java.util.LinkedList;
 
 public class MimeTypeDetector implements TypeDetector {
-    private static final Logger LOGGER = LoggerFactory.getLogger(RDFAnalyzer.class);
+	private static final Logger LOGGER = LoggerFactory.getLogger(RDFAnalyzer.class);
 
     public Lang detectMimeType(File data) {
 
+        LinkedList<FiniteStateMachine> machinesList = new LinkedList<FiniteStateMachine>(); //3.i
         Lang mimeType = null;
+
         try {
             FileInputStream inputStream = new FileInputStream(data.getAbsolutePath());
-            FiniteStateMachine machine1 = RdfAutomata.buildRDFStateMachine();
-            FiniteStateMachine machine2 = TurtleAutomata.buildTurtleStateMachine();
+
+            machinesList.add(FiniteStateMachineFactory.create(RDFLanguages.RDFXML));
+            machinesList.add(FiniteStateMachineFactory.create(RDFLanguages.TURTLE));
+
             char current;
-            while (inputStream.available() > 0 && (!machine1.canStop() || !machine2.canStop())) {
+
+            while (inputStream.available() > 0 && machinesList.size() > 1) { //3.iv (until only one machine is in list)
                 current = (char) inputStream.read();
-                if (!machine1.isError()) //whichever leads to error state is processed anymore.
-                    machine1 = machine1.switchState(String.valueOf(current));
-                if (!machine2.isError())
-                    machine2 = machine2.switchState(String.valueOf(current));
+
+                FiniteStateMachine machine = machinesList.remove(); // (machines gets removed from list for processing)
+                machine.switchState(String.valueOf(current)); //3.ii
+
+                if (!machine.isError()) // (the machine is added only if its not in error state)
+                    machinesList.addLast(machine); //3.iii
             }
 
-            if (machine1.canStop() && !machine1.isError())
-                mimeType = machine1.getMimeType();
-            if (machine2.canStop() && !machine2.isError())
-                mimeType = machine2.getMimeType();
+            mimeType = machinesList.get(0).getMimeType();
```

Comment:
You assume that your list won't be empty. In your current (faulty) implementation this might be working. However, in a correct implementation you can not assume that this is always the case. So you should prepare to handle cases in which all machines can not handle the input and the `machineList` is empty at that point.